frontend/src/metabase-lib/query.ts

```diff
@@ -72,3 +72,7 @@ export function replaceClause(
 ): Query {
   return ML.replace_clause(query, stageIndex, targetClause, newClause);
 }
+
+export function sourceTableOrCardId(query: Query): number | string | null {
```

Comment:
There's actually a `TableId` typescript type defined, you can get it by importing `metabase-types/api`. It's effectively `number | string` so you could do `TableId | null` here maybe. Not sure if that's worth it or not but it does maybe make it easier to track down the places we're using `card__<id>` "table" IDs, hopefully one day we can get rid of that icky hack entirely.